test/e2e/reconciler/apiexportendpointslice/apiexportendpointslice_test.go

```diff
@@ -352,12 +356,28 @@ func TestAPIExportEndpointSliceWithPartitionPrivate(t *testing.T) {
 
 	sliceWithAllName := sliceWithAll.Name
 
+	t.Logf("Getting the total number of shards in the system")
+	shards, err := shardClient.Cluster(core.RootCluster.Path()).List(ctx, metav1.ListOptions{})
+	require.NoError(t, err, "error listing Shards")
+
 	framework.Eventually(t, func() (bool, string) {
 		sliceWithAll, err := kcpClusterClient.Cluster(partitionClusterPath).ApisV1alpha1().APIExportEndpointSlices().Get(ctx, sliceWithAllName, metav1.GetOptions{})
 		require.NoError(t, err)
-		if len(sliceWithAll.Status.APIExportEndpoints) == 1 {
+		if len(sliceWithAll.Status.APIExportEndpoints) == len(shards.Items) {
+			for _, apiExportEndpoint := range sliceWithAll.Status.APIExportEndpoints {
+				found := false
+				for _, shard := range shards.Items {
+					if strings.Contains(apiExportEndpoint.URL, shard.Spec.VirtualWorkspaceURL) {
+						found = true
+						break
+					}
+				}
+				if !found {
+					return false, fmt.Sprintf("didn't find matching shard for %v endpoint, shards: %#v", apiExportEndpoint.URL, shards.Items)
+				}
+			}
 			return true, ""
 		}
-		return false, fmt.Sprintf("expected 1 endpoint, but got: %#v", sliceWithAll.Status.APIExportEndpoints)
+		return false, fmt.Sprintf("expected %d endpoint, but got: %#v", len(shards.Items), sliceWithAll.Status.APIExportEndpoints)
```

Comment:
I am concerned that you may be introducing a flake here: What does guarantee no shard was created by a parallel test case between your call to list shards and the one to get the APIExportEndpointSlices?

Reply:
good point, assuming schedulable shards won't be removed we could change this logic to expect `>=` `initial shards+ shards added by this tests`.